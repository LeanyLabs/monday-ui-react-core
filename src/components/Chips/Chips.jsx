import React, { useRef, forwardRef, useMemo, useCallback } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import Icon from "../Icon/Icon";
import useMergeRefs from "../../hooks/useMergeRefs";
import CloseSmall from "../Icon/Icons/components/CloseSmall";
import { getCSSVar } from "../../services/themes";
import { NOOP } from "../../utils/function-utils";
import { elementColorsNames, getElementColor } from "../../general-stories/colors/colors-vars-map";
import "./Chips.scss";

const Chips = forwardRef(
  ({ className, id, label, leftIcon, rightIcon, disabled, readOnly, color, iconSize, onDelete }, ref) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });

    const backgroundColorStyle = useMemo(() => {
      return { backgroundColor: disabled ? getCSSVar("disabled-background-color") : getElementColor(color, true) };
    }, [disabled, color]);

    const onDeleteCallback = useCallback(() => {
      if (onDelete) {
        onDelete(id);
      }
    }, [id, onDelete]);

    const hasCloseButton = !readOnly && !disabled;

    return (
      <div
        ref={mergedRef}
        className={cx("chips--wrapper", className, { disabled, "with-close": hasCloseButton })}
        id={id}
        style={backgroundColorStyle}
      >
        {leftIcon ? (
          <Icon
            className="chip-icon left"
            iconType={Icon.type.ICON_FONT}
            clickable={false}
            icon={leftIcon}
            iconSize={iconSize}
            ignoreFocusStyle
          />
        ) : null}
        <div className="label">{label}</div>
        {rightIcon ? (
          <Icon
            className="chip-icon right"
            iconType={Icon.type.ICON_FONT}
            clickable={false}
            icon={rightIcon}
            iconSize={iconSize}
            ignoreFocusStyle
          />
        ) : null}
        {hasCloseButton && (
          <Icon
            aria-label={`Remove ${label}`}
            className="chip-icon close"
            iconType={Icon.type.SVG}
            clickable
            icon={CloseSmall}
            iconSize={18}
            onClick={onDeleteCallback}
          />
        )}
      </div>
    );
  }
);

Chips.colors = elementColorsNames;

Chips.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,
  /** Icon to place on the right */
  rightIcon: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  /** Icon to place on the left */
  leftIcon: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  color: PropTypes.oneOf(Object.keys(Chips.colors)),
  /** size for font icon */
  iconSize: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  onDelete: PropTypes.func
};
Chips.defaultProps = {
  className: "",
  id: "",
  label: "",
  disabled: false,
  readOnly: false,
  rightIcon: null,
  leftIcon: null,
  color: Chips.colors.PRIMARY,
  iconSize: 16,
  onDelete: NOOP
};

export default Chips;
