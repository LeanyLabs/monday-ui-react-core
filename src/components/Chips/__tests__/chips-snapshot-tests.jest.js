import React from "react";
import Chips from "../Chips";
import renderer from "react-test-renderer";
import { Calendar } from "../../Icon/Icons";

describe("Chips renders correctly", () => {
  it("renders correctly with empty props", () => {
    const tree = renderer.create(<Chips />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with color", () => {
    const tree = renderer.create(<Chips color={Chips.colors.NEGATIVE} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  it("renders correctly disabled chip", () => {
    const tree = renderer.create(<Chips disabled={true} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly without close button", () => {
    const tree = renderer.create(<Chips readOnly={true} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with text", () => {
    const tree = renderer.create(<Chips label="text" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it.skip("renders correctly with right icon ", () => {
    const tree = renderer.create(<Chips rightIcon={Calendar} readOnly />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
