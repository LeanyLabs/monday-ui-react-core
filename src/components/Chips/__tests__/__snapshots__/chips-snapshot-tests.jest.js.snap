// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chips renders correctly renders correctly disabled chip 1`] = `
<div
  className="chips--wrapper disabled"
  id=""
  style={
    Object {
      "backgroundColor": undefined,
    }
  }
>
  <div
    className="label"
  >
    
  </div>
</div>
`;

exports[`Chips renders correctly renders correctly with color 1`] = `
<div
  className="chips--wrapper with-close"
  id=""
  style={
    Object {
      "backgroundColor": "var(--negative-color-selected)",
    }
  }
>
  <div
    className="label"
  >
    
  </div>
  <svg
    aria-hidden={false}
    className="icon_component chip-icon close icon_component--clickable"
    fill="currentColor"
    height="18"
    onClick={[Function]}
    role="button"
    tabIndex={0}
    viewBox="0 0 20 20"
    width="18"
  >
    <path
      clipRule="evenodd"
      d="M6.53033 5.46967C6.23744 5.17678 5.76256 5.17678 5.46967 5.46967C5.17678 5.76256 5.17678 6.23744 5.46967 6.53033L8.62562 9.68628L5.47045 12.8415C5.17756 13.1343 5.17756 13.6092 5.47045 13.9021C5.76334 14.195 6.23822 14.195 6.53111 13.9021L9.68628 10.7469L12.8415 13.9021C13.1343 14.195 13.6092 14.195 13.9021 13.9021C14.195 13.6092 14.195 13.1343 13.9021 12.8415L10.7469 9.68628L13.9029 6.53033C14.1958 6.23744 14.1958 5.76256 13.9029 5.46967C13.61 5.17678 13.1351 5.17678 12.8422 5.46967L9.68628 8.62562L6.53033 5.46967Z"
      fill="currentColor"
      fillRule="evenodd"
    />
  </svg>
</div>
`;

exports[`Chips renders correctly renders correctly with empty props 1`] = `
<div
  className="chips--wrapper with-close"
  id=""
  style={
    Object {
      "backgroundColor": "var(--primary-selected-color)",
    }
  }
>
  <div
    className="label"
  >
    
  </div>
  <svg
    aria-hidden={false}
    className="icon_component chip-icon close icon_component--clickable"
    fill="currentColor"
    height="18"
    onClick={[Function]}
    role="button"
    tabIndex={0}
    viewBox="0 0 20 20"
    width="18"
  >
    <path
      clipRule="evenodd"
      d="M6.53033 5.46967C6.23744 5.17678 5.76256 5.17678 5.46967 5.46967C5.17678 5.76256 5.17678 6.23744 5.46967 6.53033L8.62562 9.68628L5.47045 12.8415C5.17756 13.1343 5.17756 13.6092 5.47045 13.9021C5.76334 14.195 6.23822 14.195 6.53111 13.9021L9.68628 10.7469L12.8415 13.9021C13.1343 14.195 13.6092 14.195 13.9021 13.9021C14.195 13.6092 14.195 13.1343 13.9021 12.8415L10.7469 9.68628L13.9029 6.53033C14.1958 6.23744 14.1958 5.76256 13.9029 5.46967C13.61 5.17678 13.1351 5.17678 12.8422 5.46967L9.68628 8.62562L6.53033 5.46967Z"
      fill="currentColor"
      fillRule="evenodd"
    />
  </svg>
</div>
`;

exports[`Chips renders correctly renders correctly with text 1`] = `
<div
  className="chips--wrapper with-close"
  id=""
  style={
    Object {
      "backgroundColor": "var(--primary-selected-color)",
    }
  }
>
  <div
    className="label"
  >
    text
  </div>
  <svg
    aria-hidden={false}
    className="icon_component chip-icon close icon_component--clickable"
    fill="currentColor"
    height="18"
    onClick={[Function]}
    role="button"
    tabIndex={0}
    viewBox="0 0 20 20"
    width="18"
  >
    <path
      clipRule="evenodd"
      d="M6.53033 5.46967C6.23744 5.17678 5.76256 5.17678 5.46967 5.46967C5.17678 5.76256 5.17678 6.23744 5.46967 6.53033L8.62562 9.68628L5.47045 12.8415C5.17756 13.1343 5.17756 13.6092 5.47045 13.9021C5.76334 14.195 6.23822 14.195 6.53111 13.9021L9.68628 10.7469L12.8415 13.9021C13.1343 14.195 13.6092 14.195 13.9021 13.9021C14.195 13.6092 14.195 13.1343 13.9021 12.8415L10.7469 9.68628L13.9029 6.53033C14.1958 6.23744 14.1958 5.76256 13.9029 5.46967C13.61 5.17678 13.1351 5.17678 12.8422 5.46967L9.68628 8.62562L6.53033 5.46967Z"
      fill="currentColor"
      fillRule="evenodd"
    />
  </svg>
</div>
`;

exports[`Chips renders correctly renders correctly without close button 1`] = `
<div
  className="chips--wrapper"
  id=""
  style={
    Object {
      "backgroundColor": "var(--primary-selected-color)",
    }
  }
>
  <div
    className="label"
  >
    
  </div>
</div>
`;
